### ISSUE TITLE
layer2 mode doesn't receive broadcast packets on VM unless promiscuous mode is enabled

### ISSUE THREAD

**Issue by michaelfig** [metallb/metallb#253]
**Is this a bug report or a feature request?**: Feature request, for documentation.

**What happened**:
I couldn't connect to my LoadBalancer IP from outside the host, and the documentation provided no hints.

**What you expected to happen**:
Traffic would flow correctly, or some kind of hint would be present in the documentation.

**How to reproduce it (as minimally and precisely as possible)**:
I installed the stable/metallb Helm chart with the following resulting configmap:

```
$ kubectl get cm metallb -oyaml
apiVersion: v1
data:
  config: |
    address-pools:
    - addresses:
      - 192.168.1.39/32
      name: default
      protocol: layer2
kind: ConfigMap
metadata:
  creationTimestamp: 2018-04-26T18:22:43Z
  labels:
    app: metallb
    chart: metallb-0.5.0
    heritage: Tiller
    release: metallb
  name: metallb
  namespace: default
  resourceVersion: "3517970"
  selfLink: /api/v1/namespaces/default/configmaps/metallb
  uid: cfb29a64-497e-11e8-a2c2-6c3be52d32a5
$ kubectl get svc | grep 192.168.1
ingress-nginx-ingress-controller        LoadBalancer   192.168.223.82    192.168.1.39   80:31455/TCP,443:30129/TCP   2h
```

The node host's vmbr0 interface is 192.168.1.40/24.  I have a second host using 192.168.1.41/24, that I expect to be able to take over the loadbalancer IPs.

When I do "nc -v -z 192.168.1.39 80" from another host, it just hangs.  However, when I run 'tcpdump' on the node, the netcats and HTTP connections magically start working.  This is because tcpdump puts the interface into promiscuous mode.

I solved this by putting the following into my Debian host's /etc/network/interfaces script, under the configuration for the interface in question:

```
iface vmbr0 inet static
       [...]
       up /bin/ip link set vmbr0 promisc on
```

**Environment**:
- MetalLB version: 0.6.1
- Kubernetes version: 1.9.6
- BGP router type/version: none (using layer2)
- OS (e.g. from /etc/os-release): Debian GNU/Linux 9 (stretch)
- Kernel (e.g. `uname -a`): Linux pve2 4.13.16-1-pve #1 SMP PVE 4.13.16-43 (Fri, 16 Mar 2018 19:41:43 +0100) x86_64 GNU/Linux

**Comment by danderson** [comment]
This is interesting. This is the first time I've heard of any setup that requires promiscuous mode to make MetalLB layer2 mode work. It specifically does not rely on anything other than broadcast ethernet packets, which all correct ethernet NICs and drivers should make visible to the OS.

Based on your kernel version, it looks like you're running on Proxmox. My dev environment is on proxmox, with a bunch of clusters that work fine without the need for promiscuous mode... So there is definitely _some_ way to make it work, now we just need to figure out what's different between your proxmox and mine :)

What version of PVE are you running? Can you paste the output of `ip addr` on the PVE machine (not the VM) ? What kind of NIC hardware do you have configured on the VM?

**Comment by danderson** [comment]
Thanks for the info.

My hunch is that this is a weirdness of how linux implements bridges. Either it's filtering ARP requests based on L3 layer knowledge, or it's filtering inbound traffic that doesn't match local IPs (but that makes no sense, it would break IP forwarding... Just to be paranoid, does `cat /proc/sys/net/ipv4/ip_forward` on the PVE machine output 1?)

Once the IP is working with promisc on, if you turn promisc off again, does it keep working for a few minutes, or does it stop working immediately? When it's broken, on the client side can you run `tcpdump -i any arp or host 192.168.1.39` and try to curl or netcat again? I want to see if it's the ARP traffic or the TCP traffic that's getting blackholed.

I'll try to reproduce this locally, but my available time to work on MetalLB right now is very limited, so I can't promise quick progress :(

**Comment by michaelfig** [comment]
So, ip_forward is indeed 1.

When I turn off promisc, it does still work to 'nc -v -z 192.168.1.39'.  In fact, I can't get it to fail now.

I'll try again tomorrow from the office where I was able to get it to fail before.

Will keep you updated.

**Comment by michael-robbins** [comment]
As referenced in #284 I got this as well on my Raspberry Pi cluster running Metal LB, all nodes using wlan0 to connect to the local LAN for connectivity (inter node and rest of network).

All nodes within the cluster correctly resolved the IP and I was able to see the nginx landing page from the layer 2 tutorial.

But PC's on the local network were not able to do the same, no ARP requests were responded to by Metal LB.

By enabling promiscuous mode on wlan, all computers on the LAN were able to correctly see the nginx landing page.

Here's my log:
```
{"caller":"announcer.go:89","event":"createARPResponder","interface":"cni0","msg":"created ARP responder for interface","ts":"2018-12-24T05:21:18.138825423Z"}
{"caller":"announcer.go:94","error":"creating NDP responder for \"cni0\": listen ip6:ipv6-icmp fe80::8ce0:7ff:fe8a:4301%cni0: bind: invalid argument","interface":"cni0","msg":"failed to create NDP responder","op":"createNDPResponder","ts":"2018-12-24T05:21:18.139929008Z"}
{"caller":"announcer.go:94","error":"creating NDP responder for \"cni0\": listen ip6:ipv6-icmp fe80::8ce0:7ff:fe8a:4301%cni0: bind: invalid argument","interface":"cni0","msg":"failed to create NDP responder","op":"createNDPResponder","ts":"2018-12-24T05:21:28.142298761Z"}
{"caller":"announcer.go:98","event":"createNDPResponder","interface":"cni0","msg":"created NDP responder for interface","ts":"2018-12-24T05:21:38.147381253Z"}
{"caller":"announcer.go:89","event":"createARPResponder","interface":"veth24075984","msg":"created ARP responder for interface","ts":"2018-12-24T05:21:38.219083376Z"} {"caller":"announcer.go:98","event":"createNDPResponder","interface":"veth24075984","msg":"created NDP responder for interface","ts":"2018-12-24T05:21:38.219997379Z"} {"caller":"main.go:159","event":"startUpdate","msg":"start of service update","service":"default/nginx","ts":"2018-12-24T05:22:10.985143393Z"}
{"caller":"main.go:229","event":"serviceAnnounced","ip":"192.168.58.170","msg":"service has IP, announcing","pool":"dal-prahran-k8s-lb","protocol":"layer2","service":"default/nginx","ts":"2018-12-24T05:22:10.985626723Z"}
{"caller":"main.go:231","event":"endUpdate","msg":"end of service update","service":"default/nginx","ts":"2018-12-24T05:22:10.985969949Z"}

# curl 192.168.58.170 from PC times out
# Enabled wlan0 promiscuous mode

{"caller":"arp.go:102","interface":"wlan0","ip":"192.168.58.170","msg":"got ARP request for service IP, sending response","responseMAC":"b8:27:eb:e9:f3:6f","senderIP":"192.168.58.107","senderMAC":"78:24:af:be:54:e5","ts":"2018-12-24T05:43:49.764529457Z"}
```

The odd thing is MetalLB doesn't say it's started an ARP Responder on wlan0 initially?

**Comment by russellb** [comment]
I'm going to close this out as stale since it hasn't been updated in a couple of years.

**Comment by arthurcgc** [comment]
I can confirm that this is still an issue on my Raspberry Pi 4 model B

**Comment by virgil9306** [comment]
I am encountering this issue as well on k3s nodes running on Proxmox. 

Thus, I don't think this is a Pi-specific issue. I'm not using a Pi. 

More, recent reports are available here: https://github.com/metallb/metallb/issues/284

> What I found was that upon an initial loadbalancer being created I could access the public ip until a timeout occurred and then the service would become unavailable. If I edited the service the public IP would become available again over the wifi until that same timeout occurs. So I plugged the Pi into ethernet on the same network and boom it immediately worked and the service doesn't go down.

The same issue happens for me. I am trying to test with the nodes' NICs to promiscuous to see if that will change anything.

---

If this is a known issue, even if it's outside the scope of MetalLB, it's something that many people seem to have encountered and could be added to documentation to help people who are stuck.

I don't know the root cause of this problem, however, so can't say "in XYZ circumstances, this is how to fix it."

**Comment by bbockelm** [comment]
@virgil9306 - we tripped over this locally as well.  Here's the most succinct description of the problem I can come up with:

> When Kubernetes uses a bridge device in default configurations, the bridge device will not pass IP traffic for load-balanced IPs assigned by metallb UNLESS the pod happens to be running on the same host as the L2 IP.

Reported workarounds:
- Set `bridge-nf-call-iptables` to zero (as reported in https://github.com/metallb/metallb/issues/535; haven't verified myself).
- Set the bridge device to promiscuous mode (verified locally).

I believe this was recently re-triggered by https://github.com/metallb/metallb/pull/976 in metallb 0.11.0 because, with that code merged, the requirement that the load balanced IP be on the same host as a pod was relaxed.

(Fun story - this one was a pain to debug because the cluster would magically start working whenever I did a `tcpdump` of the interface to see where the packet was getting dropped.  One side effect of `tcpdump`, of course, is to set promiscuous mode on the bridge ... so debugging the problem "fixed" it!)

Hopefully this summary + workaround is helpful for whoever the next person hits this issue.
